File: fib2.c
#include <stdio.h>
 
int fib(int n)
{
	int k=n/2;
	if(n<2) return 1;
	int f0=fib(k);
	int f1=fib(k-1);
	if(n%2 == 0) 
		return f0*f0+f1*f1;
	else
		return f0*f0+2*f0*f1;
}

int main()
{
	printf("%d\n", fib(45));
	return 0;
}

File: fib2.pl
fib(0, 1).
fib(1, 1).
fib(N, F) :-
	N > 1,
	K is N div 2,
	K1 is K-1,
	fib(K, F0),
	fib(K1, F1),
	(   N mod 2 =:= 0
	->  F is F0*F0+F1*F1
	;   F is F0*F0+2*F0*F1), !.

